### PR TITLE
Auto-resolve certain result mismatches

### PR DESCRIPTION
**Pull request by caithagoras**
The PR extends `FailureResolver` to not only be able to resolve query failures, but also result mismatches. Added 2 types of result mismatch resolvers.

## StructuredColumnMismatchResolver
If array element or map key/value contains floating point types,
column checksum is unlikely to match. Hence, we can allow Verifier
to auto resolve those cases.
* For an array column, resolve if the element type contains floating
  point types and the cardinality matches.
* For a map column, resolve if either key or value contains floating
  point types, the cardinality sum matches, and the checksum of the
  key or value that does not contains floating point type matches.
* Resolve a test case only when all columns are resolved.

## IgnoredFunctionsMismatchResolver
In the case of a results mismatch, if the query uses a function in a
specified list, the test case is marked as resolved.
This is useful to mark certain FB functions that is known to produce
different results on Verifier.

```
== RELEASE NOTES ==

Verifier Changes
* Add support to auto-resolve result mismatch for structured-type columns.
* Add support to auto-resolve result mismatch in case the query uses functions to be ignored.
```

### CODE REVIEW

**Comment by caithagoras** [comment]
@bhhari Could you help reviewing this PR? Thanks!

File: presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ArrayColumnValidator.java [hunk 1]
```diff
@@ -53,25 +106,19 @@
...
-    public List<ColumnMatchResult> validate(Column column, ChecksumResult controlResult, ChecksumResult testResult)
+    public List<ColumnMatchResult<ArrayColumnChecksum>> validate(Column column, ChecksumResult controlResult, ChecksumResult testResult)
+    {
+        ArrayColumnChecksum controlChecksum = toColumnChecksum(column, controlResult);
+        ArrayColumnChecksum testChecksum = toColumnChecksum(column, testResult);
+
+        return ImmutableList.of(new ColumnMatchResult<>(Objects.equals(controlChecksum, testChecksum), column, controlChecksum, testChecksum));
+    }
+
+    private static ArrayColumnChecksum toColumnChecksum(Column column, ChecksumResult checksumResult)
```
Comment (yingsu00) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ArrayColumnValidator.java:117]: Should we move this private method to the private method section below?

File: presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ChecksumValidator.java [hunk 2]
```diff
@@ -55,14 +53,11 @@ public Query generateChecksumQuery(QualifiedName tableName, List<Column> columns
         return simpleQuery(new Select(false, selectItems.build()), new Table(tableName));
     }
 
-    public Map<Column, ColumnMatchResult> getMismatchedColumns(List<Column> columns, ChecksumResult controlChecksum, ChecksumResult testChecksum)
+    public List<ColumnMatchResult<?>> getMismatchedColumns(List<Column> columns, ChecksumResult controlChecksum, ChecksumResult testChecksum)
```
Comment (yingsu00) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ChecksumValidator.java:56]: Nit: I'd move this to a separate commit, but it's up to you.

File: presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ArrayColumnValidator.java [hunk 3]
```diff
@@ -37,6 +39,57 @@
...
+        @Override
+        public int hashCode()
+        {
+            return Objects.hash(checksum, cardinalitySum);
+        }
+
+        @Override
+        public String toString()
+        {
+            return format("checksum: %s, cardinality_sum: %s", checksum, cardinalitySum);
```
Comment (yingsu00) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ArrayColumnValidator.java:89]: Is `cardinality_sum` really sum of the cardinalities?

File: presto-docs/src/main/sphinx/admin/verifier.rst [hunk 4]
```diff
@@ -168,6 +168,17 @@ query failures.
 * ``COMPILER_ERROR``: Resolves if checksum fails with this error. If a control query has too many
   columns, generated checksum query might be too large in certain cases.
 
+In cases of result mismatches, Verifier may be giving noisy signals, and we allow Verifier to
+automatically resolve certain mismatches.
+
+* **Structured-typed Columns**: If array element or map key/value contains floating point types, column checksum is unlikely to match.
+    * For an array column, resolve if the element type contains floating point types and the
+      cardinality matches.
```
Comment (yingsu00) [inline presto-docs/src/main/sphinx/admin/verifier.rst:176]: cardinality checksum?

File: presto-docs/src/main/sphinx/admin/verifier.rst [hunk 5]
```diff
@@ -168,6 +168,17 @@ query failures.
...
+    * For a map column, resolve if either key or value contains floating point types, the
+      cardinality sum matches, and the checksum of the key or value that does not contains
+      floating point type matches.
```
Comment (yingsu00) [inline presto-docs/src/main/sphinx/admin/verifier.rst:179]: I feel this is a bit hard to understand. Can we say
```
For a map column, resolve the mismatch when both of the following conditions are true:
 - The cardinality checksum matches;
 - The checksum of the key or value that does not contains floating point type matches.
```

File: presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ArrayColumnValidator.java [hunk 6]
```diff
@@ -37,6 +39,57 @@
 public class ArrayColumnValidator
         implements ColumnValidator
 {
+    public static class ArrayColumnChecksum
+            extends ColumnChecksum
+    {
+        private final Object checksum;
+        private final Object cardinalitySum;
```
[the remark this answers is not in the document]
Comment (caithagoras) [reply presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ArrayColumnValidator.java]: We were not doing cardinality checksum, but that sound like a good idea to me. Will add array/map cardinality checksum for better confidence.

File: presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ArrayColumnValidator.java [hunk 7]
```diff
@@ -37,6 +39,57 @@
 public class ArrayColumnValidator
         implements ColumnValidator
 {
+    public static class ArrayColumnChecksum
```
[the remark this answers is not in the document]
Comment (caithagoras) [reply presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ArrayColumnValidator.java]: Moved to standalone class.

File: presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ColumnChecksum.java [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.verifier.checksum;
+
+public abstract class ColumnChecksum
```
[the remark this answers is not in the document]
Comment (caithagoras) [reply presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ColumnChecksum.java:16]: It has to be an abstract class if we want to force subclasses to implement those methods. There are a couple of examples in the code base. You can search for `public abstract boolean equals`

File: presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ChecksumValidator.java (hunk 2 above)
Comment (caithagoras) [reply presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ChecksumValidator.java:56]: Done

**Comment by caithagoras** [comment]
@yingsu00 Comments addressed. Thanks!

**Comment by yingsu00** [comment]
Nit: 
For the `Auto-resolve mismatched strucutured-type columns` commit, typo `strucutured` -> `structured`

> * For a map column, resolve if either key or value contains floating
  point types, the cardinality sum matches, and the checksum of the
  key or value that does not contains floating point type matches.

Can we update this commit message the same as the comment in the rst code?

File: presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ArrayColumnChecksum.java [hunk 9]
```diff
@@ -23,11 +23,13 @@
         extends StructureColumnChecksum
 {
     private final Object checksum;
+    private final Object cardinalityChecksum;
     private final Object cardinalitySum;
 
-    public ArrayColumnChecksum(@Nullable Object checksum, @Nullable Object cardinalitySum)
+    public ArrayColumnChecksum(@Nullable Object checksum, @Nullable Object cardinalityChecksum, @Nullable Object cardinalitySum)
```
[the remark this answers is not in the document]
Comment (caithagoras) [reply presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ArrayColumnChecksum.java:29]: `cardinalityChecksum` will be null when there is no input row.
`cardinalitySum` should never be null since we added a coalesce. Will update this.

**Comment by caithagoras** [comment]
@yingsu00 comments addressed.

**Comment by caithagoras** [comment]
@mbasmanova Would you mind taking a final look? Thanks!

**Comment by mbasmanova** [comment]
> @mbasmanova Would you mind taking a final look? Thanks!

@caithagoras I'll take a look later today. @yingsu00 Are you happy with the code now? Do you have any further comments? If not, please, approve this PR.

**Comment by yingsu00** [comment]
LGTM

**Comment by caithagoras** [comment]
@mbasmanova Although that might be a powerful feature to have , I think it's overly complicated mainly from the implementation perspective, and also a bit from user perspective to discern the results.

Current, we only recursively checksum sub-fields for row type, but array and map are still terminals. If we recursively checksum those structure types, the checksum query might be too complicated.

For example: We generate 4 columns for floating point types. To flatten a simple `array<double> a` column, we need 4 reduces in addition to the cardinality sum:
```
sum(reduce(a, 0, (s, x)->s+x, s->s)) a.element$sum,
sum(reduce(a, 0, (s, x)->s+if(isnan(x), 1, 0), s->s)) a.element$nan_count
sum(reduce(a, 0, (s, x)->s+if(x = infinity(), 1, 0), s->s)) a.element$positive_infinity_count
sum(reduce(a, 0, (s, x)->s+if(x = -infinity, 1, 0), s->s)) a.element$negative_infinity_count
```

Now if the structure type is slightly more complex, like `map<int, row(array<varchar>, row(double))>`, the checksum query would be disaster.

Also, it might be confusing that sometimes we flatten while sometimes we do not. For example, for `array<row(array<varchar>, array<double>)>`, we don't need to flatten `a.element[0]`, since it is a `array<array<varchar>>`, which does not contains floating point, while we need to flatten `a.element[1]`, which is a array<array<double>>.

The majority of the cases this commit intend to address is false alerts for columns like `map<string, real>`. In those cases, we found key checksums and cardinality checksums are matching, and we usually just mark the failure as resolved since it does not give high-enough signal for regression. Hence, this auto-resolution is better than none.

I would say maybe supporting recursively validating array/map columns do not have a high enough as of now, we can think deeper about the problem if that have change.

**Comment by mbasmanova** [comment]
@caithagoras Thanks for detailed explanation. Makes sense to me.
